### PR TITLE
TST Adds gradcheck and gradgradcheck to module info

### PR DESCRIPTION
**Pull request by thomasjpfan**
Follow up to #61935

cc @albanD @mruberry @jbschlosser @walterddr

### CODE REVIEW

File: test/test_modules.py [hunk 1]
```diff
@@ -141,6 +141,41 @@ def test_pickle(self, device, dtype, module_info):
...
+            m.to(device).to(dtype)
+
+            params = tuple(m.parameters())
+
+            # === Perform gradient check on the input_args ===
+            input_args, input_kwargs = module_input.forward_input.args, module_input.forward_input.kwargs
+
+            for obj in input_kwargs.values():
+                # All tensors that require gradients should be in input_args
+                self.assertFalse(isinstance(obj, torch.Tensor) and obj.requires_grad)
```
Comment (thomasjpfan) [inline test/test_modules.py:167]: I had a version that supported tensors with gradients in `input_kwargs`, but I think it was a little too complicated. I am open to supporting tensors with gradients in `input_kwargs` if it is needed.

Comment (jbschlosser) [reply test/test_modules.py:167]: I think this is fine for now, as long as it's clear to devs writing test logic that tensors with gradients in `input_kwargs` are not supported. To that end, I'd suggest replacing the assert with a call to `self.fail()` with a nice message (imo helps distinguish between the test failing because of some actual bug vs. invalid test inputs).

File: torch/testing/_internal/common_modules.py [hunk 2]
```diff
@@ -175,13 +176,14 @@ def module_inputs_torch_nn_Linear(module_info, device, dtype, requires_grad, **k
...
 
     cases: List[Tuple[str, dict]] = [
         ('', {}),
         ('ignore_index', {'ignore_index': 2}),
-        ('weights', {'weight': make_input(10)}),
-        ('weights_ignore_index', {'weight': make_input(10), 'ignore_index': 2}),
-        ('weights_ignore_index_neg', {'weight': make_input(10), 'ignore_index': -1})
+        ('weights', {'weight': make_weight(10).abs()}),
+        ('weights_ignore_index', {'weight': make_weight(10).abs(), 'ignore_index': 2}),
+        ('weights_ignore_index_neg', {'weight': make_weight(10).abs(), 'ignore_index': -1})
```
Comment (jbschlosser) [inline torch/testing/_internal/common_modules.py:186]: Are these `abs()` required due to #64256 ?

File: test/test_modules.py (hunk 1 above)
Comment (jbschlosser) [reply test/test_modules.py:167]: It might be the case that we do need to handle `input_kwargs`, if we find this limitation is too much of a barrier to expressiveness

**Review by jbschlosser — COMMENTED** [review]
Thanks for the update! I wonder if we should support any of the gradcheck-related flags that OpInfo does? https://github.com/pytorch/pytorch/blob/257623da39c9ecc63025e90a418852ef3200b57f/torch/testing/_internal/common_methods_invocations.py#L504-L523

File: torch/testing/_internal/common_modules.py (hunk 2 above)
Comment (thomasjpfan) [reply torch/testing/_internal/common_modules.py:186]: Yup. Although... is there a use case for negative weights? 🤔

**Comment by thomasjpfan** [comment]
> I wonder if we should support any of the gradcheck-related flags that OpInfo does? 

I would add the ones that are currently being used in `NewModuleTest`, which are `check_batched_grad`, `check_gradgrad` and `gradcheck_fast_mode`. The rest of the flags can be added when a `ModuleInfo` future requires them?

**Comment by thomasjpfan** [comment]
Looking at the modules, options in `common_nn.py`, I think only `supports_gradgrad` needs to be added.

Only `TransformerEncoderLayer` requires `check_batched_grad` which passes now.

The only module that uses `gradcheck_fast_mode` is `LayerNorm` for one of its samples:

https://github.com/pytorch/pytorch/blob/e4ff14ad5955f7c4d052aa44069c77654e8b5f2e/torch/testing/_internal/common_nn.py#L1635-L1644

I think this was done to speed up the gradcheck for this one specific sample.

File: torch/testing/_internal/common_modules.py (hunk 2 above)
Comment (albanD) [reply torch/testing/_internal/common_modules.py:195]: Don't we prefer to have at least one test with negative weights and an explicit skip for the gradgrad for it?
That way we make sure that the gradcheck works and it is clear that the gradgrad is problematic.

**Review by albanD — COMMENTED** [review]
Out of curiosity, what is the runtime impact of adding this?

**Comment by thomasjpfan** [comment]
> Out of curiosity, what is the runtime impact of adding this?

There are not many `ModuleInfo` added to the database, so it doesn't not add too much runtime. Currently, we are testing grad+gradgrad twice, once in `test_modules` and another in `common_nn`. In the end, we want to switch over to all module testing to `test_modules`. I envision this process:

1. Copy tests from `common_nn` into `test_modules`
2. Copy modules from `common_nn` into `common_modules` that are edge cases. i.e. `Embedding`
3. New C++ / python parity checker that looks at `common_modules`
4. After we have `test_modules` contain all the test in `common_nn`, we can have PRs removing items in the `common_nn` database and adding them into `common_modules`. This should not increase runtime because we are  removing modules from the `common_nn` database and `test_modules` are running the same tests.

File: torch/testing/_internal/common_modules.py (hunk 2 above)
Comment (jbschlosser) [reply torch/testing/_internal/common_modules.py:196]: Yeah, I wonder if we want to revisit the idea of separating metadata for samples that can be used at test collection time from the sample values so we can skip and run things on a per-sample basis. It'd be awesome to have a separate named test for each sample.

For this scenario, if we had separate tests for gradcheck and gradgradcheck, we could then skip either of them for a given sample (although it's probably not likely we'd want to skip gradcheck and run gradgradcheck).

**Review by jbschlosser — APPROVED** [review]
LGTM! I'm going to put a little more thought into how we can get per-sample skips but we can tweak that later if a solution comes around

**Comment by malfet** [comment]
Reverting, as it caused ParallelTBB failures, see here for example: https://hud.pytorch.org/commit/67ebde56459557199b3c907b81b3c819f77500b9

File: torch/testing/_internal/common_modules.py (hunk 2 above)
Comment (albanD) [reply torch/testing/_internal/common_modules.py:196]: The thing that we can use also is just split the ModuleInfo in two: the one with negative weights skips gradgrad and the one with positive weights only runs it.

File: test/test_modules.py [hunk 3]
```diff
@@ -1,12 +1,13 @@
 from inspect import signature
 from copy import deepcopy
 import tempfile
+from functools import wraps
```
Comment (jbschlosser) [inline test/test_modules.py:4]: unused, breaking lint check :(

File: torch/testing/_internal/common_utils.py [hunk 4]
```diff
@@ -2853,3 +2853,21 @@ def wrapper(*args, **kwargs):
...
+    """
+    if not IS_TBB:
+        return fn
+
+    @wraps(fn)
+    def wrap_fn(self, *args, **kwargs):
+        num_threads = torch.get_num_threads()
+        torch.set_num_threads(1)
+        output = fn(self, *args, **kwargs)
+        torch.set_num_threads(num_threads)
```
Comment (jbschlosser) [inline torch/testing/_internal/common_utils.py:2871]: should the logic for setting back to the original number of threads be in a finally clause (or context manager) to ensure it gets called even if an exception occurs?

File: torch/testing/_internal/common_modules.py [hunk 5]
```diff
@@ -66,7 +73,7 @@ def _parametrize_test(self, test, generic_cls, device_cls):
                 _update_param_kwargs(param_kwargs, 'dtype', dtype)
 
                 try:
-                    active_decorators = []
+                    active_decorators = [set_single_threaded_if_parallel_tbb]
```
Comment (jbschlosser) [inline torch/testing/_internal/common_modules.py:76]: is it a bad idea to just decorate the gradcheck tests with this instead of putting it here?

File: test/test_modules.py [hunk 6]
```diff
@@ -206,6 +206,58 @@ def test_check_inplace(self, device, dtype, module_info):
...
+
+
+    @modules(module_db, allowed_dtypes=[torch.double])
+    def test_grad(self, device, dtype, module_info):
+        self._test_gradients_helper(device, dtype, module_info, gradcheck)
+
+    @modules(module_db, allowed_dtypes=[torch.double])
+    def test_gradgrad(self, device, dtype, module_info):
+        if not module_info.supports_gradgrad:
+            self.skipTest("Skipped! Module does not support gradgrad")
```
Comment (jbschlosser) [inline test/test_modules.py:257]: A cool alternative would avoid generating the test in the first place:
```
@modules([m for m in module_db if m.supports_gradgrad], allowed_dtypes=[torch.double])
```

**Review by jbschlosser — COMMENTED** [review]
almost ready :))

**Comment by thomasjpfan** [comment]
@jbschlosser Was ParallelTBB removed from CI? 

XREF: https://github.com/pytorch/pytorch/pull/65092

File: torch/testing/_internal/common_modules.py (hunk 5 above)
Comment (thomasjpfan) [reply torch/testing/_internal/common_modules.py:76]: From reading the surrounding issues around parallel tbb, I think it can also break with the forward pass even though we have not ran across it yet.

Although, maybe all this parallel_tbb code can be removed if the tests are not running anymore? #65092

**Comment by jbschlosser** [comment]
> @jbschlosser Was ParallelTBB removed from CI?
> 
> XREF: #65092

Good find, I just asked and it's planned to be added back in at some point (once it's fixed).

File: torch/testing/_internal/common_modules.py (hunk 5 above)
Comment (jbschlosser) [reply torch/testing/_internal/common_modules.py:76]: Cool, makes sense. Do you mind adding a comment above this line explaining a bit why the decorator is here? Also, FYI I asked & ParallelTBB will be coming back at some point, so I guess we still need this :/
